### PR TITLE
Allow ui_test style headers in compiletest

### PR DESCRIPTION
**Pull request by asquared31415**
r? @oli-obk 

This is a step towards the MCP to replace compiletest https://github.com/rust-lang/compiler-team/issues/536

This PR allows compiletest to accept certain headers that ui_test can understand.  Tests that don't use any compiletest-specific features can now be run under ui_test and compiletest.

The second commit changes all the tests to use the new directives.

### CODE REVIEW

File: src/tools/compiletest/src/errors.rs
Comment (oli-obk) [inline src/tools/compiletest/src/errors.rs:120]:
```suggestion
    // ui-test style revisions are `~[revs]`
```

File: tests/ui/asm/aarch64/type-check-3.rs
Comment (oli-obk) [inline tests/ui/asm/aarch64/type-check-3.rs:1]: These haven't gotten converted

File: tests/ui/asm/aarch64/srcloc.rs
Comment (oli-obk) [inline tests/ui/asm/aarch64/srcloc.rs:2]: this isn't supported yet in ui-test

File: tests/ui/asm/aarch64/srcloc.rs
Comment (oli-obk) [inline tests/ui/asm/aarch64/srcloc.rs:3]: These haven't gotten converted yet

File: tests/ui/asm/inline-syntax.rs
Comment (oli-obk) [inline tests/ui/asm/inline-syntax.rs:3]: I think the one space after `]` is allowed to stay, but it doesn't matter much in either direction

File: tests/ui/asm/x86_64/const.rs
Comment (oli-obk) [inline tests/ui/asm/x86_64/const.rs:2]: These are called `//@run` in ui-test

File: tests/ui/asm/x86_64/issue-96797.rs
Comment (oli-obk) [inline tests/ui/asm/x86_64/issue-96797.rs:1]: maybe ui-test should have a way to register custom commands that can be converted to other commands. This command is just changing compiler flags, and only really something the rustc test suite cares about

File: tests/ui/asm/x86_64/issue-96797.rs
Comment (oli-obk) [inline tests/ui/asm/x86_64/issue-96797.rs:3]: fun. Not sure how to do this yet

File: tests/ui/associated-type-bounds/associated-item-through-where-clause.rs
Comment (oli-obk) [inline tests/ui/associated-type-bounds/associated-item-through-where-clause.rs:1]: I don't think ui-test supports this

File: tests/ui/borrowck/copy-suggestion-region-vid.fixed
Comment (oli-obk) [inline tests/ui/borrowck/copy-suggestion-region-vid.fixed:1]: stry .fixed file

**Review by oli-obk — CHANGES_REQUESTED** [review]
I reviewed up to the point where github wanted me to expand the files.

I guess the only thing is that I'd have preferred the compiletest changes to be in a few individual commits, but I reviewed the entire diff on compiletest itself and it lgtm.

File: tests/ui/asm/inline-syntax.rs
Comment (asquared31415) [reply tests/ui/asm/inline-syntax.rs:3]: Oh yep, any amount of whitespace (such that it would be removed by `trim_start`) is allowed. Since we're converting the entire test suite, this is a chance to effectively set the style that people should be using, since people are going to copy paste.  Should we do `//@[rev] command` with a single space then?

Comment (asquared31415) [reply tests/ui/asm/inline-syntax.rs:3]: Additionally we could normalize to `//@[rev] edition: 2021` such that there's a space after the `:` in a name-value directive?

Comment (oli-obk) [reply tests/ui/asm/inline-syntax.rs:3]: I like both of these

File: tests/ui/asm/x86_64/const.rs
Comment (asquared31415) [reply tests/ui/asm/x86_64/const.rs:2]: That was something I meant to ask you about and then totally forgot, thanks.

File: tests/ui/associated-type-bounds/associated-item-through-where-clause.rs
Comment (asquared31415) [reply tests/ui/associated-type-bounds/associated-item-through-where-clause.rs:1]: According to https://github.com/oli-obk/ui_test/blob/f03c5de5185f2026c65f690f74b1629505eec65b/src/parser.rs#L502 I think it does? Unless it means something different in compiletest and ui_test.

File: tests/ui/asm/aarch64/srcloc.rs
Comment (asquared31415) [reply tests/ui/asm/aarch64/srcloc.rs:3]: Thanks, missed that.

File: tests/ui/borrowck/copy-suggestion-region-vid.fixed
Comment (asquared31415) [reply tests/ui/borrowck/copy-suggestion-region-vid.fixed:1]: This actually was always supposed to be there.  The test was **always** using a `//@run-rustfix` but now that compiletest understands that, it created the fixed file (when i blessed).

File: tests/ui/asm/aarch64/type-check-3.rs
Comment (asquared31415) [reply tests/ui/asm/aarch64/type-check-3.rs:1]: I hadn't yet implemented the `only` and `ignore` things, looking into that now.

File: tests/ui/asm/x86_64/issue-96797.rs
Comment (asquared31415) [reply tests/ui/asm/x86_64/issue-96797.rs]: Well you're in luck, most of these got removed :D

(though yes it's still a general issue for migration)

**Comment by apiraino** [comment]
ping @asquared31415 for a rebase, apart from that who else should have a look. This PR maybe also needs a sign-off from:

r?  t-boostrap
r? t-libs

thanks

**Comment by asquared31415** [comment]
@rustbot author

Sorry this was not ready for review, there's still lots to do.

**Pull request closed without merging**
